mmark.1
.\" Generated by Mmark Markdown Processer - mmark.miek.nl
.TH "MMARK" 1 "April 2019" "User Commands" "Mmark Markdown"

.SH "NAME"
.PP
mmark \- generate XML, HTML or manual pages from mmark markdown documents.

.SH "SYNOPSIS"
.PP
\fBmmark\fP [\fBOPTIONS\fP] [\fIFILE...\fP]

.SH "DESCRIPTION"
.PP
\fBMmark\fP is a powerful markdown processor written in Go, geared towards writing IETF documents. It
is, however, \fIalso\fP suited for writing complete books and other technical documentation.

.PP
Mmark provides an advanced markdown dialect that processes file(s) to produce internet\-drafts in XML
RFC 7991
\[la]https://tools.ietf.org/html/rfc7991\[ra] format. Mmark can produce xml2rfc (aforementioned
RFC 7991), HTML5 and manual pages.

.PP
The syntax is detailed at https://mmark.miek.nl/syntax
\[la]https://mmark.miek.nl/syntax\[ra].

.PP
Compared to other markdown variants mmark adds:

.IP \(bu 4
(Extended) title block to specify authors and IETF specific bits.
.IP \(bu 4
(Special) sections for abstracts, prefaces or notes.
.IP \(bu 4
Including other files with the option to specify line ranges, regular expressions and/or
prefixing each line with a custom string.
.IP \(bu 4
Document divisions: front matter, main matter and back matter.
.IP \(bu 4
Captions for code, tables, quotes and subfigures.
.IP \(bu 4
Asides for small notes.
.IP \(bu 4
Figures and subfigures that allow for grouping images into subfigures as well as giving a single
image metadata (a link, attributes, etc.).
.IP \(bu 4
Block Level Attributes that allow to specify attributes, classes and IDs for elements.
.IP \(bu 4
Indices to mark an item (and/or a subitem) to be referenced in the document index.
.IP \(bu 4
Citations and adding XML References (those used by the IETF).
.IP \(bu 4
Cross references: short form of referencing a section in the document.
.IP \(bu 4
Super\- and subscript.
.IP \(bu 4
Callouts in code and text.
.IP \(bu 4
BCP14 (RFC 2119) keyword detection.


.SH "RFC 7991"
.PP
This is the XML format used by the RFC editor for accepting Internet\-Drafts.

.SH "HTML5"
.PP
The HTML5 renderer outputs HTML.

.SH "MANUAL PAGES"
.PP
The man renderer outputs nroff that can be viewed via man(1).

.SH "OPTIONS"
.TP
\fB\fC-ast\fR
print abstract syntax tree and exit
.TP
\fB\fC-fragment\fR
don't create a full document
.TP
\fB\fC-css\fR \fIURL\fP
\fIURL\fP to a CSS stylesheet (only used with \-html)
.TP
\fB\fC-head\fR \fIURL\fP
\fIURL\fP to HTML to be included in head (only used with \-html)
.TP
\fB\fC-html\fR
create HTML output
.TP
\fB\fC-man\fR
output nroff (manual pages)
.TP
\fB\fC-unsafe\fR
allow includes from anywhere in the filesystem, otherwise they are only allowed \fIbelow\fP the
current document
.TP
\fB\fC-unicode\fR
allow unicode characters in <t>, from xml2rfc 3.16.0 this is allowed. It's true by default, set
to false if you have an older xml2rfc version. Note the <u> text/html output produced by
xml2rfc is (extremely) verbose.
.TP
\fB\fC-index\fR
generate an index at the end of the document (default true)
.TP
\fB\fC-bibliography\fR
generate a bibliography section after the back matter (default true), this \fIneeds\fP a
\fB\fC{{backmatter}}\fR in the document
.TP
\fB\fC-version\fR
show mmark's version


.SH "ALSO SEE"
.PP
RFC 7991 and (maybe) RFC 7749. The main site for Mmark is
https://mmark.miek.nl
\[la]https://mmark.miek.nl\[ra]. The syntax used by mmark is explained in the syntax
page
\[la]https://mmark.miek.nl/post/syntax/\[ra].



mmark.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"

	"github.com/gomarkdown/markdown"
	"github.com/gomarkdown/markdown/ast"
	"github.com/gomarkdown/markdown/html"
	"github.com/gomarkdown/markdown/parser"
	"github.com/mmarkdown/mmark/v2/lang"
	"github.com/mmarkdown/mmark/v2/mast"
	"github.com/mmarkdown/mmark/v2/mparser"
	"github.com/mmarkdown/mmark/v2/render/man"
	"github.com/mmarkdown/mmark/v2/render/mhtml"
	"github.com/mmarkdown/mmark/v2/render/pod"
	"github.com/mmarkdown/mmark/v2/render/xml"
)

var (
	flagCSS       = flag.String("css", "", "link to a CSS stylesheet (only used with -html)")
	flagHead      = flag.String("head", "", "link to HTML to be included in head (only used with -html)")
	flagAst       = flag.Bool("ast", false, "print abstract syntax tree and exit")
	flagBib       = flag.Bool("bibliography", true, "generate a bibliography section after the back matter")
	flagFragment  = flag.Bool("fragment", false, "don't create a full document")
	flagHTML      = flag.Bool("html", false, "create HTML output")
	flagIndex     = flag.Bool("index", true, "generate an index at the end of the document")
	flagMan       = flag.Bool("man", false, "generate manual pages (nroff)")
	flagPod       = flag.Bool("pod", false, "generate Perl POD")
	flagUnsafe    = flag.Bool("unsafe", false, "allow unsafe includes")
	flagIntraEmph = flag.Bool("intra-emphasis", false, "interpret camel_case_value as emphasizing \"case\" (legacy behavior)")
	flagVersion   = flag.Bool("version", false, "show mmark version")
	flagUnicode   = flag.Bool("unicode", true, "from xml2rfc 3.16 onwards unicode is allowed in <t>")
)

func main() {
	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(), "SYNOPSIS: %s [OPTIONS] %s\n", os.Args[0], "[FILE...]")
		fmt.Println("\nOPTIONS:")
		flag.PrintDefaults()
	}

	flag.Parse()
	args := flag.Args()
	if len(args) == 0 {
		args = []string{"os.Stdin"}
	}
	if *flagVersion {
		fmt.Println(Version)
		os.Exit(0)
	}

	for _, fileName := range args {
		var (
			d    []byte
			err  error
			init mparser.Initial
		)
		if fileName == "os.Stdin" {
			init = mparser.NewInitial("")
			d, err = ioutil.ReadAll(os.Stdin)
			if err != nil {
				log.Printf("Couldn't read %q: %q", fileName, err)
				continue
			}
		} else {
			init = mparser.NewInitial(fileName)
			d, err = ioutil.ReadFile(fileName)
			if err != nil {
				log.Printf("Couldn't open %q: %q", fileName, err)
				continue
			}
		}

		d = markdown.NormalizeNewlines(d)

		if *flagUnsafe {
			init.Flags |= mparser.UnsafeInclude
		}

		if !*flagIntraEmph {
			mparser.Extensions |= parser.NoIntraEmphasis
		}

		p := parser.NewWithExtensions(mparser.Extensions)
		parserFlags := parser.FlagsNone
		documentTitle := ""      // hack to get document title from toml title block and then set it here.
		documentLanguage := "en" // get document language from title block if it is set.
		if !*flagHTML && !*flagMan {
			parserFlags |= parser.SkipFootnoteList // both xml formats don't deal with footnotes well.
		}
		p.Opts = parser.Options{
			ParserHook: func(data []byte) (ast.Node, []byte, int) {
				node, data, consumed := mparser.Hook(data)
				if t, ok := node.(*mast.Title); ok {
					documentTitle = t.TitleData.Title
					documentLanguage = t.TitleData.Language
				}
				return node, data, consumed
			},
			ReadIncludeFn: init.ReadInclude,
			Flags:         parserFlags,
		}

		doc := markdown.Parse(d, p)
		if *flagMan {
			title := false
			// If there isn't a title block the resulting manual page does not start
			// with .TH, this messes up the entire rendering. Walk to AST to check for
			// a title block, and if none is found inject an empty one.
			ast.WalkFunc(doc, func(node ast.Node, entering bool) ast.WalkStatus {
				if _, ok := node.(*mast.Title); ok {
					title = true
					return ast.Terminate
				}
				return ast.GoToNext
			})
			if !title {
				t := &mast.Title{TitleData: &mast.TitleData{Title: "User Commands 1"}}
				c := doc.GetChildren()
				newc := append([]ast.Node{t}, c...)
				doc.SetChildren(newc) // t must be the first element.
			} else {
				ast.AppendChild(doc, &mast.Authors{})
			}

		}
		if *flagBib {
			mparser.AddBibliography(doc)
		}
		if *flagIndex {
			mparser.AddIndex(doc)
		}

		if *flagAst {
			ast.Print(os.Stdout, doc)
			fmt.Print("\n")
			return
		}

		var renderer markdown.Renderer

		switch {
		case *flagHTML:
			mhtmlOpts := mhtml.RendererOptions{
				Language: lang.New(documentLanguage),
			}
			opts := html.RendererOptions{
				Comments:       [][]byte{[]byte("//"), []byte("#")}, // TODO(miek): make this an option.
				RenderNodeHook: mhtmlOpts.RenderHook,
				Flags:          html.CommonFlags | html.FootnoteNoHRTag | html.FootnoteReturnLinks,
				Generator:      `  <meta name="GENERATOR" content="github.com/mmarkdown/mmark Mmark Markdown Processor - mmark.miek.nl`,
			}
			if !*flagFragment {
				opts.Flags |= html.CompletePage
			}
			opts.CSS = *flagCSS
			if *flagHead != "" {
				head, err := ioutil.ReadFile(*flagHead)
				if err != nil {
					log.Printf("Couldn't open %q, error: %q", *flagHead, err)
					continue
				}
				opts.Head = head
			}
			if documentTitle != "" {
				opts.Title = documentTitle
			}

			renderer = html.NewRenderer(opts)
		case *flagMan:
			opts := man.RendererOptions{
				Comments: [][]byte{[]byte("//"), []byte("#")},
				Language: lang.New(documentLanguage),
			}
			if *flagFragment {
				opts.Flags |= man.ManFragment
			}
			renderer = man.NewRenderer(opts)
		case *flagPod:
			opts := pod.RendererOptions{
				Comments: [][]byte{[]byte("//"), []byte("#")},
				Language: lang.New(documentLanguage),
			}
			if *flagFragment {
				opts.Flags |= pod.PodFragment
			}
			renderer = pod.NewRenderer(opts)
		default:
			opts := xml.RendererOptions{
				Flags:    xml.CommonFlags,
				Comments: [][]byte{[]byte("//"), []byte("#")},
				Language: lang.New(documentLanguage),
			}
			if *flagFragment {
				opts.Flags |= xml.XMLFragment
			}
			if *flagUnicode {
				opts.Flags |= xml.AllowUnicode
			}

			renderer = xml.NewRenderer(opts)
		}

		x := markdown.Render(doc, renderer)

		fmt.Println(string(x))
	}
}
